### ISSUE TITLE
Android 6: new permission system

### ISSUE THREAD

**Issue by tobiasKaminsky** [owncloud/android#1343]
I have installed owncloud on a fresh android 6 system.
After installing I could not download any file because the "storage" permission was not granted.
This should, according to the new permission system, catched during runtime and the user should be asked if one wants to grant this permission.

**Comment by AndyScherzinger** [comment]
We can implement this via the AppCompat lib so it works in a backwards compatible way :) What suprises me a little is the fact that the storage permission is missing since that would lead to the same problem on any Android version and Android 6 still supports the now legacy permission model?!

ContextCompat should be able to handle all this: https://developer.android.com/training/permissions/requesting.html But since I haven't played around with this I have no feeling how easy this will work and if using ApPCompat means it will work completely backward compatible... Using this then means #1220 has to be merged first since permissions require AppCompat >= 23.x

**Comment by AndyScherzinger** [comment]
I deployed a permissions requsting implementation for WRITE_EXTERNAL_STORAGE in the Marshmallow branch and the beta version :grinning: 
This also means I implemented the new permissions API introduced with Android M in a backwards compatible (yet untested) way. Relevant commit can be checked out here: https://github.com/owncloud/android/commit/373235225780503555388bc4a099f3c517a8129a
Requesting the write permission leads to also getting the read permission when the write permission is granted by the user.

Looping in @jancborchardt @masensio @davivel @przybylski and also @LukasReschke since this is just a simple implementation to get the app working again for Marshmallow compile targeting (and the beta) but this has some impact on user interaction and is also a security/permissions topic in general. The implementation basically works like dropbox, that means whenever you open up the app the first time, add your first account and open up the list (after hitting the connect button) the app will ask you for the external storage access right. It will always ask for the right if it hasn't been granted since with compiling against v23 the app simple won't let you browse the storage, download or upload files - at least not on Marshmallow. 

In Marshmallow rights can be revoked even after installing the app and this is mainly why the check has to happen over and over again. In case the permission has been denied or revoked the app now shows a snack-bar with an explanation on why the user needs to grant the right (see screenshot 1). After the user clicks OK, the permission dialog pops up the give the user the possibility to grant the right (screenshot 2). 

Screenshot 1
![device-2015-12-29-005628](https://cloud.githubusercontent.com/assets/1315170/12027501/8f0abd60-adc7-11e5-8c0e-ab2e249648b3.png)

Screenshot 2
![device-2015-12-29-005647](https://cloud.githubusercontent.com/assets/1315170/12027502/979efc5c-adc7-11e5-8d90-82d56047f9b0.png)

Looping you all in since this change has an impact on user interaction and will whenever my Marshmallow PR gets merged will also impact the way we deal with permissions. 

Imho there are thus 2 issues that need to be discussed:
* where do we check for permissions and get them granted? (right now I request it like Dropbox right at the beginning, another way would be to do it everywhere the permission is actually needed, which is nearly impossible to handle regarding background syncs but it might work for other permissions than the external storage access)
* how do we ask for the permissions after a denial/revoke of the right? (I now did it with a snackbar which is basically a non-blocking operation)

**Comment by AndyScherzinger** [comment]
Any changes to the permission API implementation should target #1220 and the ```target_marshmallow``` branch

**Comment by AndyScherzinger** [comment]
@przybylski and @tobiasKaminsky can the two of you also test drive the latest beta branch since this is quite the change (technically not LoC wise) and shouldn't be released to f-droid before anyone besides me gave it a spin :smiley:

**Comment by przybylski** [comment]
I don't have device with 6.0 so I guess I won't be a big help :( but ill have a look in the simulator

**Comment by AndyScherzinger** [comment]
@przybylski I think seeing if this (still) works on a pre-Android-M device would be quite important since I have no idea how the AppCompat implementation behaves on older Android versions handling this...

**Comment by jancborchardt** [comment]
Proposal sounds good, this should be fixed for sure. :) A few remarks:

- What does the »OK« button do? Open the popup? Is this the normal way also Dropbox does this? Shouldn’t we simply show the popup directly when a relevant action like upload/download is performed?
- If the toast stays: Change wording to »Additional permissions required to upload & download files«, and color of »OK« to white.
- Is the layout of the permissions popup standard then? Like the files icon, and the alignment of the »don’ŧ ask again« checkbox? Because at least the checkbox looks misaligned and should either be left aligned with the text above, or should be aligned so that the text of the checkbox label is left aligned with that text. ;)

**Comment by AndyScherzinger** [comment]
@jancborchardt that is basically how it is :) I'm sorry if I wasn't specific/detailed enough. 
Scenarios how the permission implementation works (which is also inline with Google's guidelines):

Scenarios:
* User opens the app and the apps checks for the permission the first time: Screenshot 1 (permission pop up is shown)
permission pop up is shown)
* User opens the app (restart) and has denied the permission earlier: Screenshot 2 (snackbar is shown with the so called "Rationale"), clicking "OK", Screenshot 1 (https://github.com/owncloud/android/commit/77e72f24ecb2a15d4a4d381563b596a61c84e4c5
* User revokes the permission in the Android settings screen and then opens the app: Screenshot 2 (snackbar is shown with the so called "Rationale"), clicking "OK", Screenshot 1 (https://github.com/owncloud/android/commit/77e72f24ecb2a15d4a4d381563b596a61c84e4c5

To answer you other questions:
* OK opens the permission popup
* snackbar has been changed according to your comment, please review since now the text is white and the action also which makes you loose the visual for the action (except for capital letters and bold)
* The permission pop up is part of Android so this always looks like that :disappointed: The checkbox is only shown in certain scenarios but that is also decided upon by Android itself not the app.

Screenshot of the snackbar now:
![device-2015-12-29-101940](https://cloud.githubusercontent.com/assets/1315170/12032197/de2fe3ae-ae15-11e5-8992-c8fa872d15f1.png)

One "problem" still exists for now:
* The user can still choose "DENY" and basically continue working with the app but nearly nothing will work. So do we want to deal with this, since we would basically have to check for persimissions and then ask for permissions and in case of denial block actions within the app. Example:
  - User opens the app the first time, denies the access write and the clicks on upload. The upload logic doesn't check for the permission yet, so it will open the upload list view but the list will be empty since the user denied access rights.

Imho the proper way would be to always check for the needed permission and then ask for it... But that also means quite some work but it would also mean the user gets asked for a certain permission in the relevant contexts. 

Any comments on this? Looking forward to the discussion since this is really a general question.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky @przybylski can you have a look at the implementation? I moved as much as I though makes sense to a PermissionUtil class and also created a colorSnackbar method in DisplayUtil to have this centralized but left anything else to the permission asking class.

@davivel Does it make sense to explicitly make the snackbar colors available to the setup.xml? ATM it is white so it probably doesn't make sence, but I can expose the values if you want me to (at least for snackbar text and action text color.

**Comment by AndyScherzinger** [comment]
@davivel can you give me a hint where the app accesses contacts? Just asking since this permission is in the manifest and on Android M I can revoke this right but I have no idea where we would need to ask for it :disappointed:

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger As this "permission" PR seems to be getting more and more complex this should be definitivly in a new PR and not in "your" marshmallow branch.
So you need to revert these changes and instead open a PR against master.
[I will try to follow up with this discussion as soon as possible].

**Comment by AndyScherzinger** [comment]
tl;dr the marhmallow branch breaks the app without having the permission api implementation. Thus permission PRs should target the marshmallow branch, be merged into this branch and after successful testing being merged into master (after the FAB branch merge though).

Unfortunately it will have to stay on this PR and the follow up PRs regarding permissions have to target the marshmallow branch. This is imho simply due to the fact that the change on the marshmallow branch (compiletarget = 23) basically breaks the permissions und thus with the change of the branch the permission implementation needs to be "updated" thus this all should be done on this branch and then the branch should be merged to master. It all has to be done here else one would have to redo the compile target update on the other permission branches which is quite some work and in order for the app to properly work all these changes need to be in place.
Thus I would vote for the process of having permission PRs targeting/based on the marshmallow branch.

Only reason to redo this branch would be to get rid of the FAB branch being the actual base for this branch.

Just my point of view though - up for discussion. Any comment from you guys @davivel @masensio what would work best for you?

**Comment by tobiasKaminsky** [comment]
You are right that the new permission "PR" would depend on this one.
But I thought it is easier to review a smaller PR.
The new permission PR can then depend on this one and once this one is merged it is very easy to merge the new one into master. But with this approach it is easier to review both PRs.

**Comment by AndyScherzinger** [comment]
The changes for marshmallow have been rebased to master - thanks @davivel for the rebase and are now available on #1473

**Comment by davivel** [comment]
I think the current solution (checking storage permission in the app start) may be good enough to get started. But probably other scenarios need to be cover.

Ideally, the permission check should be done (additionally) as later as possible, before writing in the public storage (right now, before writing a file since there is no option to choose other storage). In case of  failure, we should ask the user about the permission.

But this simple definition may easily lead to UX problems. We could ask too much to the user, if she is trying to browse and download files (though, maybe, that could be a way of make her understand how important the permission is for the app). Or we could hit the problem when the app is not visible, since the download and upload operations go on in their respective services when the user leaves the app. In this case, the request for permissions could be triggered from the fail notification, or some other point in the UI allowing manual retries (work in progress for uploads ;) ).


Another smaller detail: the Snackbar message clarifies the reason to request for the permission, but on the other hand it is adding an extra 'tap' on the screen to the sequence of interactions the user needs to do to grant the permission. Given the kind of app this is, seems pretty clear why it could want to write the external storage. Is, then, the Snackbar really necessary? Could we just show the system dialog directly?

**Comment by AndyScherzinger** [comment]
We do show the dialog directly. The snackbar is only shown in case the user denied or revoked the permission. In that case the snackbar is shown with a description/reasoning on why the user needs to grant the permission. This is also in line with the permission API and the recommendation on how to implement the new permission system :)

**Comment by davivel** [comment]
True, @AndyScherzinger . Thanks.

Then probably the only pending weakspot is, what happens if there is a long queue of pending uploads or downloads and the storage permission is removed? Is the app response *just* uncomfortable, or is it unacceptable?

**Comment by AndyScherzinger** [comment]
Hmmm...tricky case. I guess it depends on how robust the upload is implemented. It will obviously through a permission exception and I don't know how the upload implementation handles this tight now or with the work in progress reliable uploads implementation. 

If the implementation detects the permission exception and pauses/stops the upload/download and queues the files to be processed until the permission is granted we should be fine. Best would be imho in case this happens and is then detected, that we pause, trigger a notification saying we need the permission and the click on the permission would then launch reliable uploads screen launching the permission request dialog and after permission granted resumes with the upload/download.

So depending on when we release the permission implementation and reliable uploads we could probably just add this to the reliable uploads implementation with a notification triggered upon exception. Unfortunately I can only suggest it but don't really know my way around the upload/download implementation(s) so I probably wouldn't be of any real help here :( I could help with a the notification/design/ui though.

**Comment by AndyScherzinger** [comment]
@davivel this is to some extend covered by #1220 but like we discussed we should add some more checks to the code post 1.9.2 so we should probably change this issue's milestone to 1.9.3 or backlog. What do you think?

**Comment by tobiasKaminsky** [comment]
I think a "basic" android 6 support should be included in 1.9.2

**Comment by AndyScherzinger** [comment]
thanks to #1220 already merged to master basic support is included in the upcoming 1.9.2 :smiley: :+1:

**Comment by AndyScherzinger** [comment]
@davivel Any whishes how to go on here? I would suggest closing this issue since it is sheduled for 1.9.2 and is implemented/tested so far and open a new one referencing this issue and documenting any further permission checks and their location in the new issue. Fine with you?

**Comment by davivel** [comment]
We can reschedule this one, as you suggested before. 

PR#1220 is already bound to milestone 1.9.2, so we don't lose the track.

cc @rperezb

**Comment by rperezb** [comment]
thx @davivel 
looking forward to release the new uploads view!

**Comment by Fukurokudzu** [comment]
I guess instant upload suffers from this issue too. it just does'nt work if photoes are stored on sdcard.

**Comment by AndyScherzinger** [comment]
@Fukurokudzu are you using the latest version as in version 2.0.0? Since it ships the new permission implementation and if you ever opened the app on a Android 6 device you are fine. SDCard issue (in case there are any) are not related to the Android permission system in this case.

**Comment by Fukurokudzu** [comment]
Yep, I've got the latest 2.0.0 on huawei android 6. So I'll look through sdcard bugs, thanks a lot!

**Comment by AndyScherzinger** [comment]
Okay! The other issue is rather fitting so let's discuss it there :)

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger 
In Beta: UploadFileOperation:720  -> sourceFile.delete() is not possible.
sourcefile.canWrite() returns false.
Can this be related to the new permission system? How can this file be deleted?

**Comment by tobiasKaminsky** [comment]
ContextCompat.checkSelfPermission(mContext, Manifest.permission.WRITE_EXTERNAL_STORAGE)
while debugging returns 0 which says that the app is allowed to write.
But still the file is not delete-able.

**Comment by AndyScherzinger** [comment]
To be honest, I have no clue at the moment and I don't have any device with an sd card

**Comment by tobiasKaminsky** [comment]
@davivel do you have an idea? 
This can be related to #1626

**Comment by davivel** [comment]
No, not like #1626. That's a mistake I made during the uploads view refactoring, moving a temporary file that I should have removed.

I loosely remember the SD cards need special treatment from Android 5 or Android 6, and it stayed unnoticed for a time because Nexus devices didn't ship with SD card readers anymore. We should search for that information.

Right now I don't know of any case that makes this issue critical, I'd reschedule it to another release so that we have a fast bug-fixing 2.0.1 in the market.

cc @rperezb @MTRichards @cmonteroluque

**Comment by cmonteroluque** [comment]
ok, @davivel

**Comment by davivel** [comment]
Today we can say there is no problem pending.

An improvement easy to see is getting rid of the 'contacts' permission group that appears as requested by the app, but disabled by default, and that is not really needed. Is related to the permissions requested in the Manifest to access the AccountManager, but maybe not all of them are needed. With a review of the code of ```AccountUtils``` in the oc-library, we probably could remove some of those permissions and, with some luck, make the 'contacts' permission group disappear.
